### DISCUSSION TITLE
Feature Request: Use TcUnit without extending TcUnit.Fb_TestSuite

### DISCUSSION THREAD

**Discussion by RGrabichler** [tcunit/TcUnit#199] [Q&A] [answered]
![image](https://user-images.githubusercontent.com/33341896/205684980-b4bc7b7f-a281-4626-8445-1758cce06bf9.png)

I use the TcOpen Framework. 
Therefore i have the TcoContect, where my MainStation, a TcoObject runs.
The FB i want to test derives from a TcoComponent.
I created a Library for this RfidHead-Compoonent and now i want to test it.
In TcUnit i have to create a FB which extends the TcUint.Fb_TestSuite, but i cannot inherit from two different Classes.
So how can i solve this.
![image](https://user-images.githubusercontent.com/33341896/205685777-2c18f86f-4fcc-43b9-93bd-ae0737f51ee1.png)
This would be the TestSuite with the TestMethods.

Thanks for Advices

**Comment by sagatowski** [comment]
This is not a feature request. This would break the current API for TcUnit. Regarding the other question, I will transfer this to discussions.

**Reply by RGrabichler** [reply]
Is there any Workaround, how to solve this?

**Reply by RGrabichler** [reply]
I think the Solution from Peter is perfect, so no changes required! thanks Jakob (btw to tryout testing is inspired by your series :D)

**Comment by PTKu** [comment, marked as answer]
Hi,

here is a possible solution/workaround.

You will need to inherit from `TcUnit.Fb_TestSuite` block and implement `TcoCore.ITcoContext` interface

~~~
FUNCTION_BLOCK TcUnitContext EXTENDS TcUnit.Fb_TestSuite IMPLEMENTS TcoCore.ITcoContext
VAR
	_nfContext : TcoCore.TcoContextNfCompatibility;
END_VAR
~~~

TcoCore.TcoContextNfCompatibility is a class that allows to use TcOpen components in a non  TcOpen framewok application.

You will also need to implement members of `TcoCore.ITcoContext`.

![image](https://user-images.githubusercontent.com/61538034/205711145-7fcc6b43-a49e-45de-b30e-a907b0fc2823.png)

Where in all getters you will need to return values from the instance of `_nfContext`

~~~
Context := THIS^._nfContext;
~~~
~~~
EndCycleCount := THIS^._nfContext._endCycleCount;
~~~
~~~
Environment := THIS^._nfContext._settings;
~~~

~~~
Identity :=  THIS^._nfContext._Identity;
~~~

~~~
Logger := THIS^._nfContext._logger;
~~~

~~~
RealTimeClock := THIS^._nfContext._rtc;
~~~

~~~
StartCycleCount := THIS^._nfContext._startCycleCount;
~~~

If you want to perform cyclic calls on you components you will need to place the call like this:



~~~
_nfContext.VolatileOpen();

// ALL YOUR CYCLIC CALL SHOUD BE BETRWEEB THESE TWO METHODS.

_nfContext.VolatileClose();
~~~

**Reply by RGrabichler** [reply]
Thanks alot Peter, as always genious!
I thought about Interfaces, but would never been able alone to accomplish it.

![image](https://user-images.githubusercontent.com/33341896/205855723-1903c7e7-cebd-487e-a3ac-2776b50f3406.png)
Now it works like a charm! Perfect

**Reply by rruiter87** [reply]
One extra thing needs to be added:
1. Implement the  `Main()` method for the `TcUnitContext` 
2. In `Main()` call `_mainStation();`

**Comment by sagatowski** [comment]
Thanks @PTKu! You're (as always) the best.
